# Ch10: Arrays

### An Array is a Sequence

An array is a sequence of values, called elements or items.  In a string, the values are characters.  In an array, they can be any type.

There are several ways to create a new array.  The simplest is to enclose the elements in square brackets (`[ ]`).  

An array that contains no element is called an empty array.

And we can assign array values to variables.

In [1]:
[10, 20, 30, 40]
[ "hi", 2.0, 5, [10, 20] ]    # an array within another array is called nested


cheeses = ["Cheddar", "Edam", "Gouda"]
numbers = [42, 123]
empty = []

print(cheeses, " ", numbers, " ", empty)


["Cheddar", "Edam", "Gouda"] [42, 123] Any[]

In [2]:
typeof(cheeses), typeof(numbers), typeof(empty)

(Array{String,1}, Array{Int64,1}, Array{Any,1})

### Accessing Elements and Array Slices

We can access elements using indices through the bracket operator.  

- Any integer expression can be used as an index.  
- If you try to read or write an element that does not exist, you get a BoundsError.
- The keyword end points to the last index of the array.

The `∈` operator also works.

In [3]:
cheeses[1]

"Cheddar"

In [4]:
"Edam" ∈ cheeses

true

In [5]:
t = ['a', 'b', 'c', 'd', 'e', 'f'];

println(t[1:3])
println(t[3:end])

['a', 'b', 'c']
['c', 'd', 'e', 'f']


In [6]:
t[2:3] = ['x', 'y']  # updates multiple elements
print(t)

['a', 'x', 'y', 'd', 'e', 'f']

The slice operator `[:]` makes a copy of the whole array.  This is useful when you need to perform operations that modify an array but want to keep the original.

In [7]:
t_copy = t[:]
t_copy[6] = 'z'

println("t_copy:  ", t_copy)
println("t:       ", t)

t_copy:  ['a', 'x', 'y', 'd', 'e', 'z']
t:       ['a', 'x', 'y', 'd', 'e', 'f']


Although an array can contain other arrays, each of the nested arrays still counts as a single element.

In [8]:
length( ["spam", 1, ["Brie", "Roquefort", "Camembert"], [1, 2, 3]] )

4

### Traversing an Array

The most common way to traverse the elements of an array is with a `for` loop.

In [9]:
for cheese in cheeses
    println(cheese)
end

Cheddar
Edam
Gouda


In [10]:
for x in []
    println("This never executes.")
end

This works well if you only need to read the elements of the array. If you want to write or update the elements, you
need the indices. A common way to do that is to use the built-in function `eachindex`.

In [11]:
for i in eachindex(numbers)
    numbers[i] = numbers[i] + 2
end

###   Objects and Aliases

To better appreciate the concept of mutability, we first need to understand the concept of an object and its reference.  

An object is something a variable can refer to.  Let's think of it as a location in the computer memory.  An object has a value.  (Until now, we used "object" and "value" as if they were interchangabe, but now we need to be more rigorous.) 


Two or more variables can refer to the same object.  An object with more than one reference has more than one name, so we say that the object is aliased.

We can create a new variable that refers to the same object as another variable using a simple assignment statement.

In the following example, `a` refers to an object and you assign `b = a`.  Then both variables refer to the same object.  In other words, `a` and `b` are aliases for the same object (array).

In [12]:
a = [1, 2, 3]
b = a

b === a

true

### Mutable and Immutable Objects

#### Immutable Objects 

Immutable objects cannot be changed.  And hence the name.  Examples are strings and numbers.  In the following examples, Julia created only one string or number object.   `a` and `b` refer to the same object (string) and  `c` and `d` refer to the same object (number).  That is, the string has two aliases and the number has two aliases.

In [13]:
a = "banana"
b = "banana"
a === b

true

In [14]:
c = 1
d = 1
c === d

true

If we do some operation on these immutable objects, like addition and concatenation, Julia simply creates new objects that have those different values.  

The original immutable objects are still there in the computer memory.  However, if we reassign the variables with these new objects (with the new values), we  lose the references to the original objects.  

(Julia do so-called garbage collection to free up the "orphaned" memory locations).

In [15]:
a = 1
b = 1  # we crete an alias `b` so we can keep the reference to the original object (number 1)
a = a + 1  # now, `a` references a new object (number 2)
print("$a , $b , $(a === b)")

2 , 1 , false

#### Mutable Objects

Mutable objects can be changed.  **Arrays are mutable.**  That is, we can add or delete elements and we can update individual elements of arrays.

(But as we shall see later, if we reassign an existing array variable, Julia creates a new  array that happens to have the same variable name.)

In [16]:
numbers = [42, 123]
numbers[2] = 5
print(numbers)

[42, 5]

Another characteristic of mutable objects.  In the following example, even though `n` and `m` contain the same elements (and hence, equivalent, ==), they point to two different objects.  They are not identical.

In [78]:
n = [1, 2, 3] 
m = [1, 2, 3]
n === m, n == m

(false, true)

This makes sense because these are mutable objects and we don't know what their values will be later in the program.  They happen to have the same initial values.  We should really create two different objects.

**If the aliased object is mutable, changes made with one alias affect the other.**

In [18]:
a = [1, 2, 3]
b = a
b[1] = 42

print(a)

[42, 2, 3]

  
**Make copies of mutable objects to avoid aliasing:**  Although the above behavior can be useful, it is error-prone.  If you want to use a function like sort! that modifies an array (a mutable object) passed to it as the argument but also need to keep the original array as well, then you can make a copy using a slicing operation with `[:]`.

**A new assignment statement for the existing array creates a new array rather than updating that existing array.**

In [19]:
tt = [1, 2, 3]

# First, let's create an alias
ty = tt
result1 = (ty === tt)  # check to make sure they reference the same object

# and then update some of the elements.
tt[2:3] = [4,5]
result2 = (ty === tt)  #  `tt` and `ty` still reference the same object, which has been modified


# But, if we make a new assignment statement for tt, a new array that happens to have the same name `tt` is created.
# `tt` and `ty` no longer reference the same objects.
tt = [1, 4, 5]
result3 = (ty === tt)


result1, result2, result3

(true, true, false)

### Array Arguments

Let's apply what we just learned about mutable objects to function calls.

If we call a function with an argument, Julia first executes an assignment operation `parameter = argument`.  That is, the parameter becomes an alias for the object referenced by the argument.

So, when you pass an array as an argment to a function, then the parameter of the function becomes an alias of the array.  

Since these are aliases for a mutable object (array), if the function modifies the parameter, then the array (that is, the argument) also changes.  The caller sees the change.

Let's first examine the function `replace3rd`:

In [20]:
function replace3rd!(t)
    t[3] = 'r'
end

letters = ['a', 'b', 'c']
replace3rd!(letters)   # The parameter `t` and the variable `letters` are aliases for the same object!
print(letters)         # if `t` is modified then so does `letters`

['a', 'b', 'r']

In [21]:
function myreplace(t)
    t[2:end] = ['e', 'f']
end

letters = ['a', 'b', 'c']

myreplace(letters)
print(letters)

['a', 'e', 'f']

When we call `myreplace`, the parameter `t` and the variable `letters` become aliases for the same mutable object (`['a', 'b', 'c']`).  The assignment statement replaces the second to the last elements with new ones. Since `letters` also references this same object, it is also changed.

The next function `removefirst` works differently:

In [22]:
function removefirst(t)
    t = t[2:end]  # this assignment statement creates a new array that happens to have the same name `t`!  
end

letters = ['a', 'b', 'c']

letters2 = removefirst(letters);

println(letters)  # letters, which was the argument in the function call, remains unchanged
print(letters2) # here is the new array called letters2


['a', 'b', 'c']
['b', 'c']

At the beginning of `removefirst`, `t` and `letters` refer to the same array, like before.  The slice operator creates a new array and the assignment statement makes `t` refer to this new array.  But that doesn’t affect the caller. At the end, `t` refers to the new array and `letters` still refers to the original, unmodified array.

As one last set of examples, let's compare `push!`, which modifies an array by adding an element to the end of the array, and `vcat`, which creates a new array from concatenating two arrays.

In [23]:
t1 = [1, 2];
t2 = push!(t1, 3);
println(t1)  # t1 is modified and t2 is an alias for t1
println(t2)
println(t1 === t2)

t3 = vcat(t1, [4]); # t1 is not modified and a new object (array) t3 is created.
println(t1)
println(t3)
println(t1 === t3)


[1, 2, 3]
[1, 2, 3]
true
[1, 2, 3]
[1, 2, 3, 4]
false


### Array Library

Note that, as a style convention in Julia, ! is appended to names of functions that modify arrays passed as their arguments (as opposed to create new ones).

* `push!` adds a new element to the end of an array

* `pushfirst!` inserts an element at the beginningof the array.

* `insert!` inserts an element at a given index.

* `append!` add the elements of the second array to the end of the first.  **The second argument should be an array!**



* `pop!` deletes and returns the last element

* `popfirst!` deletes and returns the first element

* `splice!` delete an element of the given index and return its value

*  `deleteat!` removes an element without returing its value


* `sort!` arranges the elements of the array from low to high

* `sort`, in contrast, returns a new copy of the array with elements in order)



In [24]:
t = ['a', 'b', 'c'];
push!(t, 'd');
print(t)

['a', 'b', 'c', 'd']

In [25]:
t = ['a', 'b', 'c'];
pushfirst!(t, 'z');
print(t)

['z', 'a', 'b', 'c']

In [26]:
t = ['a', 'b', 'c'];
insert!(t, 2, 'x')
print(t)

['a', 'x', 'b', 'c']

In [27]:
t1 = ['a', 'b', 'c'];
t2 = ['d', 'e'];
append!(t1, t2);  #t2 is an array.  It remains unchanged.
println(t1)
print(t2)

['a', 'b', 'c', 'd', 'e']
['d', 'e']

In [28]:
t = ['a', 'b', 'c'];
pop!(t)

'c': ASCII/Unicode U+0063 (category Ll: Letter, lowercase)

In [29]:
print(t)

['a', 'b']

In [30]:
t = ['a', 'b', 'c'];
popfirst!(t)

'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)

In [31]:
print(t)

['b', 'c']

In [32]:
t = ['a', 'b', 'c'];
splice!(t, 2)

'b': ASCII/Unicode U+0062 (category Ll: Letter, lowercase)

In [33]:
print(t)

['a', 'c']

In [34]:
t = ['a', 'b', 'c'];
deleteat!(t, 2)

2-element Array{Char,1}:
 'a'
 'c'

In [35]:
t = ['d', 'c', 'e', 'b', 'a'];
sort!(t);
print(t)

['a', 'b', 'c', 'd', 'e']

In [36]:
t1 = ['d', 'c', 'e', 'b', 'a'];
t2 = sort(t1);  # t1 is unchanged.  a new array t2 is created.
println(t1)
print(t2)

['d', 'c', 'e', 'b', 'a']
['a', 'b', 'c', 'd', 'e']

### Accumulator, Reduce, Map, and Filter

To add up all the numbers in an array, you can use a loop like this:

In [37]:
function addall(t)
    total = 0
    for x in t
        total += x  # augmented assignment statement
    end
    total
end

addall (generic function with 1 method)

**Accumulator**: As the loop runs, total accumulates the sum of the elements; a variable used this way is sometimes called an
accumulator.  Julia offers a built-in function `sum` that works like `addall`.

In [38]:
t = [1, 2, 3, 4]
addall(t), sum(t)

(10, 10)

**Reduce Operation**: An operation like the above that combines a sequence of elements into a single value is sometimes called a reduce operation.

Often you want to traverse one array while building another. For example, the following function takes an array of strings
and returns a new array that contains capitalized strings:

In [39]:
function capitalizeall(t)
    res = []
    for s in t
        push!(res, uppercase(s))
    end
    res
end

capitalizeall (generic function with 1 method)

**Map**: An operation like `capitalizeall` is sometimes called a map because it “maps” a function (in this case uppercase ) onto each of the elements in a sequence.

Another common operation is to select some of the elements from an array and return a subarray. 

For example, the following function takes an array of strings and returns an array that contains only the uppercase strings:

In [40]:
function onlyupper(t)
    res = []        # initialize to an empty array
    for s in t
        if s == uppercase(s)
            push!(res, s)
        end
    end
    res
end

onlyupper (generic function with 1 method)

**Filter**: An operation like `onlyupper` is called a filter because it selects some of the elements and filters out the others

### Dot Syntax

For every binary operator like `^`, there is a corresponding dot operator `.^` that is automatically defined to perform `^`
element-by-element on arrays.

In [41]:
print([1, 2, 3] .^ 3)

[1, 8, 27]

Any Julia function `f` can be applied elementwise to any array with the dot syntax. 

For example to capitalize an array of strings, we don’t need an explicit loop:

In [42]:
t = uppercase.(["abc", "def", "ghi"]);
print(t)

["ABC", "DEF", "GHI"]

This is an elegant way to create a map. The function capitalizeall can be implemented by a one-liner:

In [43]:
function capitalizeall(t)
    uppercase.(t)
end

capitalizeall (generic function with 1 method)


### Arrays and Strings
A string is a sequence of characters and an array is a sequence of values, but an array of characters is not the same as a string. To convert from a string to an array of characters, you can use the function `collect`, which breaks a string or another sequence into individual elements:

In [44]:
t = collect("spam");
print(t)

['s', 'p', 'a', 'm']

If you want to break a string into words, you can use the `split` function:

In [45]:
t = split("pining for the fjords");
print(t)

SubString{String}["pining", "for", "the", "fjords"]

An optional argument called a delimiter specifies which characters to use as word boundaries.  The default delimter is the space.  The following example uses a hyphen as a delimiter:

In [46]:
t = split("spam-spam-spam", '-');
print(t)

SubString{String}["spam", "spam", "spam"]

`join` is the inverse of split . It takes an array of strings and concatenates the elements:

In [47]:
t = ["pining", "for", "the", "fjords"];
s = join(t, ' ')  # a space character is used as the delimiter.  Omit the second argument to join strings without spaces.

"pining for the fjords"

### Exercise 10-1

Write a function called `nestedsum` that takes an array of arrays of integers and adds up the elements from all of the nested arrays.

In [48]:
# function nestedsum(t)
#   sum(sum.(t))
# end

function nestedsum(t)
  total = 0
  for nested in t
    total += sum(nested)
  end
  total
end

nestedsum (generic function with 1 method)

In [49]:
t = [[1, 2], [3], [4, 5, 6]];
nestedsum(t)

21

### Exercise 10-2

Write a function called `cumulsum` that takes an array of numbers and returns the cumulative sum; that is, a new array where the th element is the sum of the first elements from the original array.

In [122]:
function cumulsum(t)
  total = 0
  res = []
  for x in t
    total += x
    push!(res, total)
  end
  res
end

cumulsum (generic function with 1 method)

In [51]:
t = [1, 2, 3];
print(cumulsum(t))

Any[1, 3, 6]

### Exercise 10-3

Write a function called `interior` that takes an array and returns a new array that contains all but the first and last elements.

In [52]:
function interior(t)
  t[2:end-1]
end

interior (generic function with 1 method)

In [53]:
t = [1, 2, 3, 4];
print(interior(t))

[2, 3]

### Exercise 10-4

Write a function called `interior!` that takes an array, modifies it by removing the first and last elements, and returns nothing.

In [54]:
function interior!(t)
  popfirst!(t)
  pop!(t)
  nothing
end

interior! (generic function with 1 method)

In [55]:
t = [1, 2, 3, 4];
interior!(t)
print(t)

[2, 3]

### Exercise 10-5

Write a function called `issort` that takes an array as a parameter and returns true if the array is sorted in ascending order and false otherwise.

In [56]:
function issort(t)
  t == sort(t)  # do they have the same values (elements) although not necessarily reference the identical object.
end

issort (generic function with 1 method)

In [57]:
issort([1, 2, 2]), issort(['b', 'a'])

(true, false)

### Exercise 10-6

Two words are anagrams if you can rearrange the letters from one to spell the other. Write a function called `isanagram` that takes two strings and returns true if they are anagrams.

TIP: You may want to first write a function that converts a string into an array of all lower case letters and no space character.

In [58]:
# function isanagram(str1,str2)
#     sort(collect(str1)) == sort(collect(str2))
# end

# Ok, there was a bug in the above implementation as it does not account for spaces and case sensitivity. 
# That was discovered through (New York Times example)

function str_to_arr(str)
    
#   res=[]
#    
#     for c in str
#         if c != ' '
#             push!(res,lowercase(c))
#         end
#     end
#    res
    
    # fancier implementation using array comprehension
    
    [ lowercase(item) for item in str if item !=' ' ]
    
     
end

        
function isanagram(str1,str2)
    
    sort(str_to_arr(str1)) == sort(str_to_arr(str2))
end



isanagram (generic function with 1 method)

In [59]:
isanagram("rail safety","fairy tales") 

true

In [60]:
isanagram("New York Times","monkeys write")

true

In [61]:
isanagram("angered","enraged")

true

### Exercise 10-7

Write a function called `hasduplicates` that takes an array and returns true if there is any element that appears more than once. It should not modify the original array.

In [62]:
function hasduplicates(arr)
    
    for index in eachindex(arr[1:end-1])
        
        if arr[index] ∈ arr[index+1:end]
             return true
        end
        
    end
    
    return false
    
end


hasduplicates (generic function with 1 method)

In [63]:
hasduplicates([1,2,3,4,5,6,7,8,9,10])

false

In [64]:
hasduplicates([1,2,3,4,5,6,7,8,9,1])

true

## Exercise 10-8

This exercise pertains to the so-called Birthday Paradox, which you can read about at https://en.wikipedia.org/wiki/Birthday_paradox.

If there are 23 students in your class, what are the chances that two (or more) of you have the same birthday? You can estimate this probability by generating random samples of 23 birthdays and checking for matches. 

TIP: You can generate random birthdays with rand(1:365).  Create an array of 23 randomly chosen birthdays and use `hasduplicates` to see if there are duplicate birthdays.  Repeat this process for many times (like 10,000 trials?) and count true trials to compute the probability.

In [65]:
count = 0
maxcount = 10000

for case in 1:maxcount
    
    birthdays=[]
    for i = 1:23
        push!(birthdays,rand(1:365))
    end
    
    if hasduplicates(birthdays)
        count += 1
    end
end

count/maxcount
    

0.5012

## Exercise 10-9

Write a function that reads the file words.txt and builds an array with one element per word. Write two versions of this function, one using push! and the other using the idiom t = [t..., x] . Which one takes longer to run? Why?

In [91]:
function word_array_builder_push()
    
    word_array = []
    
    for word in eachline("words.txt")
        push!(word_array,word)
    end

    return word_array
    
end

function word_array_builder_idiom()
    
    word_array = []
    
    for word in eachline("words.txt")
        word_array = [word_array...,word]
    end
    
    return word_array

end

word_array_builder_idiom (generic function with 1 method)

In [68]:
@time word_array_builder_idiom()

 55.533887 seconds (718.92 k allocations: 96.535 GiB, 8.43% gc time)


113809-element Array{String,1}:
 "aa"        
 "aah"       
 "aahed"     
 "aahing"    
 "aahs"      
 "aal"       
 "aalii"     
 "aaliis"    
 "aals"      
 "aardvark"  
 "aardvarks" 
 "aardwolf"  
 "aardwolves"
 ⋮           
 "zymes"     
 "zymogen"   
 "zymogene"  
 "zymogenes" 
 "zymogens"  
 "zymologies"
 "zymology"  
 "zymoses"   
 "zymosis"   
 "zymotic"   
 "zymurgies" 
 "zymurgy"   

In [92]:
@time mywords = word_array_builder_push()

  0.023449 seconds (232.96 k allocations: 7.458 MiB)


113809-element Array{Any,1}:
 "aa"        
 "aah"       
 "aahed"     
 "aahing"    
 "aahs"      
 "aal"       
 "aalii"     
 "aaliis"    
 "aals"      
 "aardvark"  
 "aardvarks" 
 "aardwolf"  
 "aardwolves"
 ⋮           
 "zymes"     
 "zymogen"   
 "zymogene"  
 "zymogenes" 
 "zymogens"  
 "zymologies"
 "zymology"  
 "zymoses"   
 "zymosis"   
 "zymotic"   
 "zymurgies" 
 "zymurgy"   

### Exercise 10-10

To check whether a word is in the word array, you could use the ∈ operator, but it would be slow because it searches through the words in order. Because the words are in alphabetical order, we can speed things up with a bisection search (also known as binary search), which is similar to what you do when you look a word up in the dictionary. You start in the middle and check to see whether the word you are looking for comes before the word in the middle of the array. If so, you search the first half of  the array the same way. Otherwise you search the second half. Either way, you cut the remaining search space in half. If the word array has 113,809 words, it will take about 17 steps to find the word or conclude that it’s not there.

Write a function called `inbisect` that takes a sorted array and a target value and returns true if the word is in the array and false if it’s not.

TIP: Maybe it is time to practice recursion?

In [93]:
function inbisect(sorted_array, target)
    
    #@show sorted_array
    a_len = length(sorted_array)
    
    if a_len < 2 
        return sorted_array[1] == target
    end

    halfind = a_len ÷ 2
    half = sorted_array[halfind]
    
    #@show a_len, halfind, half
    if half < target
        inbisect(sorted_array[(halfind+1):end],target)
    else
        inbisect(sorted_array[1:(halfind)],target)
    end
    
end
        

inbisect (generic function with 1 method)

In [96]:
ar = ["a","b","c","d","e"]
inbisect(ar,"a")

true

In [97]:
ar = ["a","b","c","d","e"]
inbisect(ar,"e")

true

In [98]:
ar = ["a","b","c","d","e"]
inbisect(ar,"bc")

false

In [99]:
ar = ["a","b","c","d","e"]
inbisect(ar,"cd")

false

In [100]:
br = ["ab","bc","cd","de","ef","fg"]
inbisect(br,"cc")

false

In [101]:
br = ["ab","bc","cd","de","ef","fg"]
inbisect(br,"dd")

false

In [102]:
inbisect(mywords,"zebra")

true

In [103]:
inbisect(mywords,"hello")

true

### Exercise 10-11

Two words are a “reverse pair” if each is the reverse of the other. Write a program `reversepairs` that finds all the reverse pairs in the word array.  Use `inbisect`.

In [104]:
function reversepairs(wordarray)
    
    for word in wordarray
        
        reverse_word = word[end:-1:1]
        if inbisect(wordarray, reverse_word)
            println(word, reverse_word)
        end
        
    end
        
end

reversepairs (generic function with 1 method)

In [105]:
reversepairs(mywords)

aaaa
abaaba
abuttuba
adda
adossoda
agaaga
agarraga
agassaga
agenessenega
ahha
ahaaha
aiderredia
airtsstria
ajarraja
alaala
aliffila
alulaalula
amma
amaama
amennema
amissima
anna
anaana
angerregna
animallamina
animessemina
annaanna
anonnona
anteetna
areera
aressera
arillira
arrissirra
arummura
atta
ateeta
atesseta
auksskua
avaava
aviddiva
avoova
awaawa
ayya
baddab
baggab
ballab
balsslab
bannab
barddrab
bassab
battab
batsstab
beddeb
benneb
bibbib
biddib
biggib
binnib
binssnib
birddrib
bissib
bobbob
boggob
boobboob
bossob
botsstob
bowsswob
braddarb
braggarb
bubbub
buddub
bunnub
bunssnub
burrub
burddrub
burggrub
bussub
buttub
butsstub
cammac
cappac
caresserac
civiccivic
coddoc
crammarc
cudduc
daad
dabbad
daddad
daggad
dahhad
dahsshad
dammad
dappad
darbbrad
darttrad
dawwad
debbed
debuttubed
decallaced
dedalladed
deeddeed
deemmeed
deeppeed
deepsspeed
deerreed
deesseed
deferrefed
degamiimaged
deifieddeified
deifierreified
deillied
dekeeked
dekeddeked
delled
deleddeled
delffled
deliverreviled


### Exercise 10-12

Two words “interlock” if taking alternating letters from each forms a new word. For example, “shoe” and “cold” interlock to form “schooled”.


Credit: This exercise is inspired by an example at http://puzzlers.org.

1. Write a program that finds all pairs of words that interlock.

TIP: Don’t enumerate all pairs and see whether they work!  That is, don't do the following as it will take forever to run.

```
for word1 in mywords
    
    for word2 in mywords

...

```

2. Can you find any words that are three-way interlocked; that is, every third letter forms a word, starting from the first,second or third?

In [108]:
list_2_interlock = []

for word in mywords
    
    if length(word) % 2 == 0
            
        index = 1
        word1 = []
        word2 = []
        
        while index < length(word)
            push!(word1,word[index])
            push!(word2,word[index+1])
            index += 2
        end
            
        if inbisect(mywords,join(word1)) && inbisect(mywords,join(word2))
            push!(list_2_interlock,word)
        end
        
    end
    
end


length(list_2_interlock)
            
            
                
                

652

In [109]:
list_2_interlock[100]

"bourne"

In [113]:
list_3_interlock = []

for word in mywords
    
    if length(word) % 3 == 0
            
        index = 1
        word1 = []
        word2 = []
        word3 = []
        
        while index < length(word)
            push!(word1,word[index])
            push!(word2,word[index+1])
            push!(word3,word[index+2])
            index += 3
        end
            
        if inbisect(mywords,join(word1)) && inbisect(mywords,join(word2)) && inbisect(mywords,join(word3))
            push!(list_3_interlock,word)
        end
        
    end
    
end


length(list_3_interlock)

348

In [118]:
list_3_interlock[200]

"manana"

In [120]:
inbisect(mywords,"na")

true